📔 jjeong - 1주차 과제 MVC(인증)

jeong724 opened the pull request:
# ✒️ 과제

## 🔑 과제 내용

1. TODO
- [x] 1단계 - 로그인
- [x] 2단계 - 로그인 리팩토링
- [x] 3단계 - 관리자 기능
<img width="1282" alt="image" src="https://github.com/user-attachments/assets/3b271552-ba6e-4716-8c51-4bf4ecb99768" />


## 🧐 고민한 내용 + 공부한 내용
우선 과제를 하면서 쿠키와 토큰 그리고 HandlerMethodArgumentResolver, HandlerInterceptor에 대해서 새롭게 알게되었습니다.. 어렵네요...
- 1단계 구현
1단계를 구현하면서 많은 걸 배웠는데, 그중에서도 쿠키에 대해 새롭게 이해하게 되었습니다. **쿠키는 서버에서 클라이언트(브라우저)로 전달되어 저장되는 작은 데이터 조각입니다.** 로그인에 성공하면 서버는 JWT 토큰을 쿠키에 담아 응답하고, 이 쿠키는 브라우저에 저장됩니다.
이후 클라이언트가 서버에 요청을 보낼 때마다 브라우저가 자동으로 쿠키를 함께 전송해주기 때문에, 사용자는 매번 로그인할 필요 없이 인증 상태를 유지할 수 있습니다. 서버는 요청에 포함된 쿠키에서 토큰을 꺼내 검증한 뒤, 유효한 토큰이라면 해당 사용자가 로그인된 상태임을 판단합니다. 
우선 1단계에서 원하는 것이 /login을 만드는 것이었기 때문에 이메일과 패스워드를 받아 로그인을 하면 jwt토큰을 쿠키에 저장하는 api를 만들었습니다. 이때 중요한건 HttpServletResponse도 함께 받아야합니다! HttpServletResponse는 스프링이 자동으로 넣어주는 서버 응답 객체로 쿠키를 만들고 응답에 추가할 때 사용합니다. `response.addCookie()`를 통해 브라우저에 쿠키를 저장합니다
예시를 보면 응답에 `Set-Cookie: token=eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxIiwibmFtZSI6ImFkbWluIiwicm9sZSI6IkFETUlOIn0.cwnHsltFeEtOzMHs2Q5-ItawgvBZ140OyWecppNlLoI; Path=/; HttpOnly` 형태로 되어있는데 이는 서버가 token이라는 쿠키를 주는 것입니다. Set-Cookie는 서버가 쿠키를 설정할 때 사용하는 HTTP 응답 헤더입니다.
저는 TokenUtil이라는 클래스를 만들어서 토큰과 관련된 메서드들을 정리했는데 이렇게 한 이유는 중복된 부분도 많고 한 곳에서 관리하면 가독성과 책임분리가 더 잘 될것이라고 생각했기 때문입니다. 
jwt토큰을 만들때 주어진 예시를 따라 멤버의 ID를 subject로 하고 추가 정보로 role과 name을 함께 주었습니다.
.
/check/login api 는 Cookie에서 토큰 정보를 추출해야하기 때문에 TokenUtil에 extractTokenFromCookie 메서드를 만들어서 브라우저가 가지고 있는 쿠키로부터 토큰을 추출해내도록 했습니다. 이때 HttpServletRequest는 브라우저가 서버로 보낸 요청 정보를 담고 있는 객체로 쿠키에 대한 정보를 가지고 있습니다.` request.getCookies();`로 쿠키를 꺼낼 수 있습니다. 쿠키에서 추출한 토큰을 통해 멤버 객체를 찾고 이를 반환합니다.
.
1단계를 구현하며 정리되지 않았던 내용들이 PR을 작성하며 점점 정리가 되고 있어서 좋네요..ㅎ 중요한 쿠키 그리고 토큰에 대해서 배울 수 있었던 단계였습니다
- 2단계 구현
HandlerMethodArgumentResolver라는 인터페이스를 이용해 구현하는 단계였습니다. HandlerMethodArgumentResolver는 처음 들어봐서 컨트롤러의 파라미터 값을 자동으로 주입해주는 인터페이스입니다. 로그인에서는 JWT 토큰에서 로그인 정보를 꺼내서 컨트롤러 파라미터에 자동으로 넣을 수 있습니다.
인터페이스에 있는 두가지 메소드를 오버라이딩 해야하는데 하나는 supportsParameter로 파라미터로 지정한 클래스가 왔을때 resolveArgument() 호출해서 객체에 값을 자동으로 넣어줍니다. 따라서 제가 작성한 resolveArgument()에서는 브라우저로부터 받은 쿠키에서 토큰을 추출해 멤버를 반환합니다.
‼️ 트러블 슈팅 
1. 저는 해당 클래스만 작성하면 된다고 생각하고 실행을 했는데 적용이 안되는 문제가 발생했습니다. 알고보니 LoginMemberArgumentResolver를 스프링에 등록하는 과정이 필요하다고 합니다.. 스프링의 웹 설정을 커스터마이징하기 위한 클래스인 WebMvcConfigurer인터페이스를 구현하는 클래스를 만들어서 `addArgumentResolvers`를 해줍니다! 이래야 잘 등록이 되어 작동합니다!
2. 아래의 조건을 보지 않고 구현해서 테스트케이스가 계속 실패했습니다.. 이 후 깨닫고 케이스를 나눠 name값을 각각 다르게 설정했습니다.
<img width="702" alt="image" src="https://github.com/user-attachments/assets/8e23c81b-aecf-48e2-b616-27b07621224b" />

- 3단계 구현
HandlerInterceptor라는 인터페이스를 이용해 구현하는 단계였습니다. 이를 통해 컨트롤러 진입 전에 쿠키에서 role을 확인하고, 어드민인 경우만 어드민 페이지가 허락되도록 하는 것이 조건입니다. HandlerInterceptor는 컨트롤러 실행 전, 실행 후, 뷰 렌더링 후에 요청을 가로채서 처리할 수 있게 해주는 인터셉터 인터페이스입니다. 이중 컨트롤러 이전에 가로채야하므로 `preHandle`을 구현했습니다.  이것도 마찬가지로 `addInterceptors`를 통해 등록이 필요합니다!


## ❓궁금한 내용
- LoginMemberArgumentResolver에서 주어진 코드에 `private MemberService memberService;` 를 의존성 주입받아서 `MemberResponse member = memberService.checkLogin(token);`을 통해 멤버 객체를 얻는게 왜 MemberDao를 의존성 주입받아서 `MemberDao member = memberDao.findById(tokenUtil.parseTokenToId(token));`을 하는 것보다 나은지 궁금합니다!

jeong724 committed:
/login api 추가

jeong724 committed:
/login/check api 추가

jeong724 committed:
/login/check api 추가

jeong724 committed:
로그인 리팩토링 : HandlerMethodArgumentResolver 활용

jeong724 committed:
로그인 리팩토링 error : memberDao is null

jeong724 committed:
로그인 리팩토링 error fix : remove memberDao -> checklogin

jeong724 committed:
handlerintercepter 구현, 스텝2 테스트 케이스 에러

jeong724 committed:
테스트2 통과하도록 fix

jeong724 committed:
Merge branch 'jjeong' into jjeong

DWL21 commented on src/main/java/com/yourssu/roomescape/utils/TokenUtil.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public String createToken(Member member){
+        String secretKey = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
```
시크릿 키를 코드에 하드코딩 하면 어떤 문제가 있을까요?

jeong724 replied:
이렇게 시크릿 키를 하드코딩할 시에 GitHub에 노출 위험이 있어 누군가 이 키를 입수하면 토큰을 위조하거나 탈취한 사용자처럼 위장할 수 있습니다. 따라서 @ConfigurationProperties로 설정 주입한 방식을 사용하도록 수정했습니다!

DWL21 commented on src/main/java/com/yourssu/roomescape/utils/TokenUtil.java:
```diff
@@ -0,0 +1,40 @@
+
+    public Long parseTokenToId(String token){
+        return Long.valueOf(Jwts.parser()
+                .verifyWith(Keys.hmacShaKeyFor("Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=".getBytes()))
```
createToken에서도 동일한 시크릿 키를 사용하고 있네요. 중복을 제거할 수 있지 않을까요?

DWL21 commented on src/main/java/com/yourssu/roomescape/utils/TokenUtil.java:
```diff
@@ -0,0 +1,40 @@
+    public String extractTokenFromCookie(Cookie[] cookies){
+        if (cookies != null) {
+            for (Cookie cookie : cookies) {
+                if (cookie.getName().equals("token")) {
```
cookies에 null이 있는 경우에는 어떻게 될까요?

jeong724 replied:
cookies 배열 안에 null이 있는 경우 null pointer exception이 발생하기 때문에 이 경우에도 예외를 처리해주도록 수정하겠습니다.

DWL21 commented on src/main/java/com/yourssu/roomescape/reservation/ReservationRequest.java:
```diff
@@ -21,4 +17,5 @@ public Long getTheme() {
     public Long getTime() {
         return time;
     }
+    public String getName(){return name;}
```
코드 정렬 기능을 사용하는게 좋아보여요~

jeong724 replied:
넵!

DWL21 commented on src/main/java/com/yourssu/roomescape/reservation/ReservationDao.java:
```diff
@@ -1,5 +1,6 @@
 package com.yourssu.roomescape.reservation;
 
+import com.yourssu.roomescape.member.LoginMember;
```
커밋 전에 사용하지 않는 import는 지워주세요.

shon5544 replied:
import문에 좌클릭하고,

윈도우 유저라면: alt + enter
맥 유저면 : option + enter

이거하면 안쓰는 import 제거 같은 옵션이 떠요. 참고해주세요~~

jeong724 replied:
오 넵! 감사합니다!!

DWL21 commented on src/main/java/com/yourssu/roomescape/reservation/ReservationDao.java:
```diff
@@ -43,12 +44,12 @@ public List<Reservation> findAll() {
     }
 
-    public Reservation save(ReservationRequest reservationRequest) {
+    public Reservation save(ReservationRequest reservationRequest, String name) {
```
ReservationRequest reservationRequest에 name이 있는데 name을 파라미터로 받을 필요가 있을까요?

jeong724 replied:
이 경우에는 과제 조건에서 
"ReservationReqeust에 name값이 있으면 name으로 Member를 찾고 없으면 로그인 정보를 활용해서 Member를 찾도록 수정합니다."는 조건이 있어서 ReservationReqeust에 name값이 있는지 없는지에 따라 다른 name을 주기 위해 이렇게 설정했습니다!

DWL21 replied:
ReservationRequest가 Controller에서 Dao까지 넘어오고 있어요. 해당 부분을 다음 과제에서 수정해보면 좋겠어요.

DWL21 commented on src/main/java/com/yourssu/roomescape/member/MemberService.java:
```diff
@@ -1,17 +1,42 @@
+    }
+
+    public String login(LoginPostRequest request){
+        Member member = memberDao.findByEmailAndPassword(request.getEmail(), request.getPassword());
+
+        if (member == null){
+            throw new IllegalArgumentException(NO_EXIST_MEMBER.getMessage());
+        }
```
member NotFound 상황을 Service Layer에서 확인하는게 맞을까요?

jeong724 replied:
컨트롤러 계층에서 검증하도록 수정했습니다!

jeong724 replied:
MemberValidator 클래스를 따로 만들어서 member NotFound 상황에서 예외를 발생시키도록 했습니다..! 
상태 없이, 단순한 검증 로직을 재사용하기 위해서 static 메서드로 구현하였고 이럴 경우 객체를 매번 생성할 필요 없이 static으로 선언해 바로 호출할 수 있다는 장점이 있습니다..
또한 LoginPostRequest를 기반으로 회원 정보를 조회하는 로직은 서비스 계층의 책임이라고 판단하여, 해당 검증 메서드는 서비스 계층에서 호출하도록 했습니다만 이 부분은 뭐가 정답일지 잘 모르겠습니다..!

DWL21 commented on src/main/java/com/yourssu/roomescape/member/MemberService.java:
```diff
@@ -1,17 +1,42 @@
 @Service
 public class MemberService {
     private MemberDao memberDao;
+    private final TokenUtil tokenUtil = new TokenUtil();
```
TokenUtil을 인스턴스로 만든 이유가 있나요? 싱글톤으로 관리해도 충분하지 않을까요?

jeong724 replied:
넵! 싱글톤으로 관리하는게 나을것같네요 수정했습니다!

DWL21 commented on src/main/java/com/yourssu/roomescape/member/MemberController.java:
```diff
@@ -32,4 +36,27 @@ public ResponseEntity logout(HttpServletResponse response) {
+            MemberResponse response = memberService.checkLogin(token);
+            return ResponseEntity.ok(response);
+        }
+        return null;
```
로그인 확인 실패 경우에 null 대신 400번대의 ResponseEntity를 반환하는게 어떤가요?

jeong724 replied:
넵! 401 에러를 반환하도록 수정했습니다!

DWL21 reviewed (changes requested):
Jjeong 1주차 과제 구현하느라 고생 많았습니다. 앞으로도 지금처럼 설계 의도에 대해 고민한 흔적을 PR에 남겨주면 좋겠습니다. 코멘트 질문에 대한 답은 Layered Architecture에 대해 공부해보면 알 수 있어요. 관련 링크 첨부합니다. 리뷰 확인하여 4/14 23:59까지 수정하여 최종 제출해주세요.
https://github.com/iluwatar/java-design-patterns
https://java-design-patterns.com/patterns/layered-architecture/#when-to-use-the-layered-architecture-pattern-in-java

jeong724 committed:
pr 코리 반영
1. token util 싱글톤 관리
2. secret key @ConfigurationProperties를 사용해 property값 주입받아 사용

jeong724 committed:
pr 코리 반영
1. MemberValidator 클래스 생성 -> 검증 책임 분리
2. 로그인 실패 시 401 에러 반환
